javascript/features/player_settings/player_settings_supplement.js

```diff
@@ -0,0 +1,117 @@
+
+    constructor(player) {
+        super();
+        this.player_ = player;
+
+        this.settings_ = new Map();
+
+        // Import the settings from the |SettingList| in to the local state.
+        for (const setting of SettingList)
+            this.settings_.set(setting.identifier, setting.clone());
+
+        // The database instance used to read and write persistent values.
+        this.database_ = server.isTest() ? new MockPlayerSettingsDatabase()
+            : new PlayerSettingsDatabase();
+    }
+
+     // Gets the map of the settings.
+     get settings() { return this.settings_; }
+
+     // Gets the map of all settings with category set as announcement
+     get announcementSettings() {
+         return new Map([...this.settings_]
+            .filter(([_, setting]) => setting.category == PlayerSetting.CATEGORY.ANNOUNCEMENT));
+     }
+
+     loadSettings() {
+        if(!this.player_.isRegistered()) {
+            return;
+        }   
+        
+        // Load the existing persistent values from the database, and apply them to the local state.
+        Promise.resolve(this.database_.loadSettings(this.player_.userId)).then(settings => {
+            for (const [identifier, value] of settings) {
+                try {
+                    this.setValue(identifier, value);
+                } catch (exception) {
+                    console.log(`Warning: Unable to restore the player setting of ${identifier}: ` +
+                        exception);
+                }
+            }
+        });
+
+     }
+     
+    // Gets the value currently assigned to the |identifier|. Will throw when the |identifier| is
+    // not known to the settings system, since we can't return anything sensible in that case.
+    getValue(identifier) {
+        const setting = this.settings_.get(identifier);
+        if (!setting)
+            throw new Error('Invalid setting given: ' + identifier);
+
+        return setting.value;
+    }
+
+    // Sets the value of the setting identified by |identifier| to |value|. Type checking will be
+    // done to make sure that the updated value is valid for the setting.
+    setValue(identifier, value) {
+        const setting = this.settings_.get(identifier);
+        if (!setting)
+            throw new Error('Invalid setting given: ' + identifier);
+
+        switch (setting.type) {
+            case PlayerSetting.TYPE_BOOLEAN:
+                if (typeof value !== 'boolean')
+                    throw new Error('The value for ' + identifier + ' must be a boolean.');
+                break;
+
+            default:
+                throw new Error('Unknown type for the setting ' + identifier + '.');
+        }
+
+        setting.value = value;
+    }
+
+    updateSetting(identifier, value) {
+        this.setValue(identifier, value);
+        
+        if(!this.player_.isRegistered()) {
+            return;
+        }
+
+        var setting = this.player_.settings.settings.get(identifier);
```

Comment:
Can we just access `this.settings_.get()` here?

Reply:
Added a getSetting so we can get the raw setting instead of only the value 👍